rados/omap.go

```diff
@@ -7,63 +7,234 @@ package rados
 import "C"
 
 import (
+	"runtime"
 	"unsafe"
 )
 
-// SetOmap appends the map `pairs` to the omap `oid`
-func (ioctx *IOContext) SetOmap(oid string, pairs map[string][]byte) error {
-	c_oid := C.CString(oid)
-	defer C.free(unsafe.Pointer(c_oid))
+var (
+	strpSize  C.size_t // the size of a char* pointer
+	sizeTSize C.size_t // the size of size_t - how recursive!
+)
 
-	var s C.size_t
+func init() {
 	var c *C.char
-	ptrSize := unsafe.Sizeof(c)
+	strpSize = C.size_t(unsafe.Sizeof(c))
+	var s C.size_t
+	sizeTSize = C.size_t(unsafe.Sizeof(s))
+}
+
+// setOmapStep is a write op step. It holds C memory used in the operation.
+type setOmapStep struct {
+	withRefs
+	withoutUpdate
+
+	// user input
+	pairs map[string][]byte
+
+	// C arguments
+	cKeys    **C.char
+	cValues  **C.char
+	cLengths *C.size_t
+	cNum     C.size_t
+}
 
-	c_keys := C.malloc(C.size_t(len(pairs)) * C.size_t(ptrSize))
-	c_values := C.malloc(C.size_t(len(pairs)) * C.size_t(ptrSize))
-	c_lengths := C.malloc(C.size_t(len(pairs)) * C.size_t(unsafe.Sizeof(s)))
+func newSetOmapStep(pairs map[string][]byte) *setOmapStep {
 
-	defer C.free(unsafe.Pointer(c_keys))
-	defer C.free(unsafe.Pointer(c_values))
-	defer C.free(unsafe.Pointer(c_lengths))
+	maplen := C.size_t(len(pairs))
+	cKeys := C.malloc(maplen * strpSize)
+	cValues := C.malloc(maplen * strpSize)
+	cLengths := C.malloc(maplen * sizeTSize)
 
+	sos := &setOmapStep{
+		pairs:    pairs,
+		cKeys:    (**C.char)(cKeys),
+		cValues:  (**C.char)(cValues),
+		cLengths: (*C.size_t)(cLengths),
+		cNum:     C.size_t(len(pairs)),
+	}
+	sos.addRefs(cKeys, cValues, cLengths)
+
+	psize := uintptr(strpSize)
 	i := 0
 	for key, value := range pairs {
 		// key
-		c_key_ptr := (**C.char)(unsafe.Pointer(uintptr(c_keys) + uintptr(i)*ptrSize))
-		*c_key_ptr = C.CString(key)
-		defer C.free(unsafe.Pointer(*c_key_ptr))
+		ck := C.CString(key)
+		sos.addRefs(unsafe.Pointer(ck))
+		ckp := (**C.char)(unsafe.Pointer(uintptr(cKeys) + uintptr(i)*psize))
+		*ckp = ck
 
 		// value and its length
-		c_value_ptr := (**C.char)(unsafe.Pointer(uintptr(c_values) + uintptr(i)*ptrSize))
-
-		var c_length C.size_t
-		if len(value) > 0 {
-			*c_value_ptr = (*C.char)(unsafe.Pointer(&value[0]))
-			c_length = C.size_t(len(value))
+		cvp := (**C.char)(unsafe.Pointer(uintptr(cValues) + uintptr(i)*psize))
+		vlen := C.size_t(len(value))
+		if vlen > 0 {
+			*cvp = (*C.char)(unsafe.Pointer(&value[0]))
```

Comment:
Ok, apparently there is `C.CBytes()` in analogy to `C.CString()`, so you should write:
```go
	cv := C.CBytes(key)
	sos.addRefs(cv)
	*cvp = cv
```

Reply:
I don't think this is a violation. If it is, its one I'm comfortable making as we have over 60 lines of code in go ceph using this technique and its the only one I know of that avoid excessive copying of data.

Additionally, I tried searching around for an official stance on this but could not find one. I did find references to the technique and none of the core go devs responded with a "don't do that":
https://github.com/golang/go/issues/14210#issuecomment-223169706
https://github.com/golang/go/issues/40636

Are some examples.

Reply:
I'm pretty sure it is a violation. Your first example passes a pointer to a C struct, and the second is about passing Go pointer to a C function, not about storing it in C memory, which is legal if the C function doesn't store it longer than it return. Here, however, we store a Go pointer in C memory for an indefinite time. The GC could decide any time to move the slice, and it will not correct the pointer to it, because it's in C memory, so it will become dangling and crash.

Reply:
From [Passing pointers](https://golang.org/cmd/cgo/#hdr-Passing_pointers)
> C code may not keep a copy of a Go pointer after the call returns.
[...]
Go code may not store a Go pointer in C memory. C code may store Go pointers in C memory, subject to the rule above: it must stop storing the Go pointer when the C function returns.

Reply:
And here is the proof:
<details><summary>(Click to expand)</summary>

<pre>
[svanders:~/golang/test] $ cat pointertest.go 
package main

// typedef void* ptr;
import "C"

import (
	"fmt"
)

func main() {
	a := []byte("This is a test")
	goPtr := &a[0]
	cPtr := (**byte)(C.malloc(C.sizeof_ptr))
	*cPtr = goPtr
	fmt.Printf("%p\n", cPtr)
}

[svanders:~/golang/test] $ GODEBUG=cgocheck=2 go run pointertest.go 
write of Go pointer 0xc0000b2010 to non-Go memory 0x65e820
fatal error: Go pointer stored into non-Go memory

runtime stack:
runtime.throw(0x4c57ce, 0x24)
	/usr/lib/golang/src/runtime/panic.go:1116 +0x72
runtime.cgoCheckWriteBarrier.func1()
	/usr/lib/golang/src/runtime/cgocheck.go:55 +0xa9
runtime.systemstack(0x462cb4)
	/usr/lib/golang/src/runtime/asm_amd64.s:370 +0x66
runtime.mstart()
	/usr/lib/golang/src/runtime/proc.go:1116

goroutine 1 [running]:
runtime.systemstack_switch()
	/usr/lib/golang/src/runtime/asm_amd64.s:330 fp=0xc00004ee20 sp=0xc00004ee18 pc=0x462de0
runtime.cgoCheckWriteBarrier(0x65e820, 0xc0000b2010)
	/usr/lib/golang/src/runtime/cgocheck.go:53 +0xcf fp=0xc00004ee58 sp=0xc00004ee20 pc=0x40588f
runtime.wbBufFlush(0x65e820, 0xc0000b2010)
	/usr/lib/golang/src/runtime/mwbbuf.go:196 +0x85 fp=0xc00004ee88 sp=0xc00004ee58 pc=0x42ee05
runtime.gcWriteBarrier(0x8, 0x65e820, 0x4431ca, 0x54c3e0, 0xc000000180, 0xc00004ef78, 0x406e45, 0xc00008e058, 0x0, 0x0, ...)
	/usr/lib/golang/src/runtime/asm_amd64.s:1461 +0xae fp=0xc00004ef10 sp=0xc00004ee88 pc=0x464c8e
main.main()
	/home/svanders/golang/test/pointertest.go:14 +0xf9 fp=0xc00004ef88 sp=0xc00004ef10 pc=0x49c3f9
runtime.main()
	/usr/lib/golang/src/runtime/proc.go:204 +0x209 fp=0xc00004efe0 sp=0xc00004ef88 pc=0x436289
runtime.goexit()
	/usr/lib/golang/src/runtime/asm_amd64.s:1374 +0x1 fp=0xc00004efe8 sp=0xc00004efe0 pc=0x464bc1
exit status 2
</pre>
</details>

Reply:
Got it. I was too fixated on the `unsafe.Pointer &b[0])` idiom, and not enough what we were *doing* with the result. Unfortunately, that means more copies in the code.
Since this is a metadata related op I think its probably fine to do copies for now. We can look into other techniques for performance in the future.

Reply:
Changed to use CBytes *big sigh*